### PR TITLE
Enable Multi DB

### PR DESCRIPTION
**Pull request by Pan-XT**
#### Why I did it
Enable Multi DB for improving routing  loading performance. The detail performancee data could be found in the following HLD
[HLD] [MultiDB HLD](https://github.com/sonic-net/SONiC/blob/master/doc/database/multi_database_instances.md)
https://github.com/sonic-net/SONiC/pull/1855

#### How I did it
modify [dockers/docker-database/Dockerfile.j2] and add [multi_database_config.json.j2]
modify [dockers/docker-database/critical_processes.j2] to resolve the formatting issues caused by multiple Redis processes.
modify [dockers/docker-database/docker-database-init.sh] to enable Multi DB if /etc/sonic/enable_multidb exists and ensure that the redis service has the necessary permissions to access /var/lib/$inst
modify [Makefile.work] [slave.mk] [/files/build_templates/sonic_debian_extension.j2 ] for setting ENABLE_MULTIDB=y when building SONiC image.

#### How I verified it
https://github.com/sonic-net/sonic-mgmt/pull/15084

### CODE REVIEW

**Comment by eddieruan-alibaba** [comment]
Double committed to Phoenixwing https://github.com/eddieruan-alibaba/sonic-buildimage/commit/9f9f4d81a181ed65d221037e55024d0869277976

and verified on vSONiC

admin@PE3:~$ ps aux | grep redis
message+    1011  1.4  0.2  86764 22372 pts/0    Sl   04:30   0:31 /usr/bin/redis-server 127.0.0.1:6379
message+    1012  0.6  0.2  77548 22316 pts/0    Sl   04:30   0:14 /usr/bin/redis-server 127.0.0.1:6380
message+    1015  0.5  0.2  64748 17696 pts/0    Sl   04:30   0:11 /usr/bin/redis-server 127.0.0.1:6381
message+    1016  1.9  0.3  71404 22532 pts/0    Sl   04:30   0:43 /usr/bin/redis-server 127.0.0.1:6382
message+    1018  0.2  0.2  62188 16456 pts/0    Sl   04:30   0:05 /usr/bin/redis-server 127.0.0.1:6383

**Comment by eddieruan-alibaba** [comment]
@StormLiangMS, @liushilongbuaa  Can you trigger the following command? 

Github Branch not ready
Please wait a few minutes to run again!
'/azpw ms_conflict'

**Review by eddieruan-alibaba — APPROVED** [review]
This change is enabled in Phoenix wing and verified in Phoenix wing's daily sanity.

http://phoenixwing.com.cn/vsonic

**Comment by eddieruan-alibaba** [comment]
Validated with phoenix wing vsonic image

admin@PE3:~$ ls -l /etc/sonic/enable_multidb 
-rw-r--r-- 1 root root 0 Oct 11 01:46 /etc/sonic/enable_multidb
admin@PE3:~$ ps aux | grep redis
message+     990  0.8  0.5  86764 39228 pts/0    Sl   06:28   7:06 /usr/bin/redis-server 127.0.0.1:6379
message+     991  0.8  0.2  71404 21248 pts/0    Rl   06:28   6:50 /usr/bin/redis-server 127.0.0.1:6380
message+     992  0.3  0.2  64748 17384 pts/0    Sl   06:28   2:28 /usr/bin/redis-server 127.0.0.1:6381
message+     994  2.9  0.3  71404 24000 pts/0    Sl   06:28  23:44 /usr/bin/redis-server 127.0.0.1:6382
message+     996  0.1  0.2  62188 16904 pts/0    Sl   06:28   1:22 /usr/bin/redis-server 127.0.0.1:6383
admin     111122  0.0  0.0   6868   648 pts/0    S+   19:51   0:00 grep redis

File: dockers/docker-database/multi_database_config.json.j2 [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            "port" : 6382,
+            "unix_socket_path" : "/var/run/redis{{DEV}}/redis3.sock",
+            "persistence_for_warm_boot" : "yes"
+            {% if DATABASE_TYPE is defined and DATABASE_TYPE != "" %}
+            ,"database_type": "{{DATABASE_TYPE}}"
+            {% endif %}
+        },
+        "redis4": {
+            "hostname" : "{{HOST_IP}}",
+            "bind": "240.1.1.2 127.0.0.1",
```
Comment (kperumalbfn) [inline dockers/docker-database/multi_database_config.json.j2:42]: @Pan-XT Why bind to 240.1.1.2?

File: dockers/docker-database/docker-database-init.sh [hunk 2]
```diff
@@ -48,7 +48,11 @@ mkdir -p /etc/supervisor/conf.d/
 if [ -f /etc/sonic/database_config$NAMESPACE_ID.json ]; then
     cp /etc/sonic/database_config$NAMESPACE_ID.json $REDIS_DIR/sonic-db/database_config.json
 else
-    HOST_IP=$host_ip REDIS_PORT=$redis_port DATABASE_TYPE=$DATABASE_TYPE j2 /usr/share/sonic/templates/database_config.json.j2 > $REDIS_DIR/sonic-db/database_config.json
+    if [ -f /etc/sonic/enable_multidb ]; then
+        HOST_IP=$host_ip REDIS_PORT=$redis_port DATABASE_TYPE=$DATABASE_TYPE j2 /usr/share/sonic/templates/multi_database_config.json.j2 > $REDIS_DIR/sonic-db/database_config.json
+    else
+        HOST_IP=$host_ip REDIS_PORT=$redis_port DATABASE_TYPE=$DATABASE_TYPE j2 /usr/share/sonic/templates/database_config.json.j2 > $REDIS_DIR/sonic-db/database_config.json
+    fi
```
Comment (kperumalbfn) [inline dockers/docker-database/docker-database-init.sh:55]: @Pan-XT  any sonic-mgmt tests for multi DB?

File: dockers/docker-database/Dockerfile.j2 [hunk 3]
```diff
@@ -48,5 +48,6 @@ COPY ["files/supervisor-proc-exit-listener", "/usr/bin"]
 COPY ["files/sysctl-net.conf", "/etc/sysctl.d/"]
 COPY ["files/update_chassisdb_config", "/usr/local/bin/"]
 COPY ["flush_unused_database", "/usr/local/bin/"]
+COPY ["multi_database_config.json.j2", "/usr/share/sonic/templates/"]
```
Comment (kperumalbfn) [inline dockers/docker-database/Dockerfile.j2:51]: @Pan-XT Could you upload a write-up for this multi-DB including the current issues and usage.

File: dockers/docker-database/multi_database_config.json.j2 (hunk 1 above)
Comment (Pan-XT) [reply dockers/docker-database/multi_database_config.json.j2:42]: 240.1.1.2 is the communication address between the Host and BMC. If the community version does not have BMC, the redis4/BMC_DB can be removed.

File: dockers/docker-database/docker-database-init.sh (hunk 2 above)
Comment (Pan-XT) [reply dockers/docker-database/docker-database-init.sh:55]: https://github.com/sonic-net/sonic-mgmt/pull/15084

File: dockers/docker-database/Dockerfile.j2 (hunk 3 above)
Comment (Pan-XT) [reply dockers/docker-database/Dockerfile.j2:51]: https://github.com/dzhangalibaba/SONiC/blob/master/doc/database/multi_database_instances.md

**Review by eddieruan-alibaba — APPROVED** [review]

**Comment by kperumalbfn** [comment]
@saiarcot895 could you review this PR

**Comment by kcudnik** [comment]
> @saiarcot895 @kcudnik could you review this PR

Can you fix build errors ?

**Comment by eddieruan-alibaba** [comment]
> > @saiarcot895 @kcudnik could you review this PR
> 
> Can you fix build errors ?

The build failure was due to yesterday's connection outage

2024-11-04T08:09:16.9633070Z get_url_version https://sonicstorage.blob.core.windows.net/public/fips/bookworm/1.4.3-1/armhf/symcrypt-openssl_1.4.3-1_armhf.deb failed
2024-11-04T08:09:16.9633607Z make[1]: *** [Makefile:12: /sonic/target/debs/bookworm/symcrypt-openssl_1.4.3-1_armhf.deb] Error 1

Could someone help to trigger a rebuild? @kperumalbfn @kcudnik

File: dockers/docker-database/multi_database_config.json.j2 [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            "port" : {{REDIS_PORT}},
+            "unix_socket_path" : "/var/run/redis{{DEV}}/redis.sock",
+            "persistence_for_warm_boot" : "yes"
+            {% if DATABASE_TYPE is defined and DATABASE_TYPE != "" %}
+            ,"database_type": "{{DATABASE_TYPE}}"
+            {% endif %}
+        },
+        "redis1": {
+           "hostname" : "{{HOST_IP}}",
+            "port" : 6380,
```
Comment (oleksandrivantsiv) [inline dockers/docker-database/multi_database_config.json.j2:15]: The port selection conflicts with both the chassis DB and Smart Switch DB service and must be changed. Port 6380 is used by the chassis Redis server, while ports 6381-6388 are assigned to the DPU databases.

**Review by oleksandrivantsiv — CHANGES_REQUESTED** [review]

File: dockers/docker-database/multi_database_config.json.j2 (hunk 4 above)
Comment (eddieruan-alibaba) [reply dockers/docker-database/multi_database_config.json.j2:15]: @oleksandrivantsiv where does chassis DB's port get defined?

This multi-db approach is enabled only on the platform which needs performance and controlled by compile flag. ENABLE_MULTIDB=y  The detail information could be found at https://lists.sonicfoundation.dev/g/sonic-wg-routing/wiki/37961

During the discussion in routing WG, we feel there is NO use case to enable it for chassis based platform for the following reasons.

1. The scale for chassis based approach is limited. MSFT uses it as T1, and Alibaba uses it as gateway. The benefit is less. 
2. Even chassis case really wants to use it. It may only wants to enable it for main switch asic, not for each DPU.  That means we need to create a new j2 for that case.  You may want to blindly create 5 redis instances for each DPU database. That causes too many redis instances in the system.

Comment (eddieruan-alibaba) [reply dockers/docker-database/multi_database_config.json.j2:15]: Discussed with Guahan on this one. 6380 will be kept for Chassis DB. It is not easy to change chassis's port number. 

We use 6379, 6378, 6377, ...  a.k.a going down for new redis instances. 

@Pan-XT can you update your diff?

Comment (Pan-XT) [reply dockers/docker-database/multi_database_config.json.j2]: Update redis1,reids2,redis3,redis4 port number to 6378,6377,6376,6375

**Review by kcudnik — APPROVED** [review]

File: dockers/docker-database/supervisord.conf.j2 [hunk 5]
```diff
@@ -31,6 +31,8 @@ dependent_startup=true
 {% if INSTANCES %}
 {%     for redis_inst, redis_items in INSTANCES.items() %}
 {%- if redis_inst != 'remote_redis' %}
+
```
Comment (liuh-80) [inline dockers/docker-database/supervisord.conf.j2:34]: Suggest remove unnecessary change in this file

File: dockers/docker-database/multi_database_config.json.j2 [hunk 6]
```diff
@@ -0,0 +1,192 @@
+{% set include_remote_db = (REMOTE_DB_IP is defined and REMOTE_DB_PORT is defined) %}
+{
+    "INSTANCES": {
+        "redis":{
+            "hostname" : "{{HOST_IP}}",
+            "port" : {{REDIS_PORT}},
+            "unix_socket_path" : "/var/run/redis{{DEV}}/redis.sock",
+            "persistence_for_warm_boot" : "yes"
+            {% if DATABASE_TYPE is defined and DATABASE_TYPE != "" %}
+            ,"database_type": "{{DATABASE_TYPE}}"
```
Comment (liuh-80) [inline dockers/docker-database/multi_database_config.json.j2:10]: I didn't find code in sonic-swss-common to parse and handle 'database_type', is there a PR?

File: dockers/docker-database/multi_database_config.json.j2 [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            "id" : 13,
+            "separator": "|",
+            "instance" : "redis_chassis"
+        }
+{% if DATABASE_TYPE is defined and DATABASE_TYPE == "dpudb" %}
+        ,
+        "DPU_APPL_DB" : {
+            "id" : 15,
+            "separator": ":",
+            "instance" : {% if include_remote_db %} "remote_redis" {% else %} "redis" {% endif %},
```
Comment (liuh-80) [inline dockers/docker-database/multi_database_config.json.j2:166]: Suggest define a variable for all these dupe template

File: dockers/docker-database/docker-database-init.sh [hunk 8]
```diff
@@ -125,6 +129,12 @@ do
     else
         echo -n > /var/lib/$inst/dump.rdb
     fi
+
+    chown -R redis:redis /var/lib/$inst
```
Comment (liuh-80) [inline dockers/docker-database/docker-database-init.sh:133]: Add comments to this file, help other understand.
Also, this seems not related with enable multi-DB, can this move to another PR?

File: dockers/docker-database/critical_processes.j2 [hunk 9]
```diff
@@ -1,5 +1,6 @@
 {% if INSTANCES %}
 {%     for redis_inst, redis_items in INSTANCES.items() %}
+
```
Comment (liuh-80) [inline dockers/docker-database/critical_processes.j2:3]: Remove this unnecessary change?

Comment (Pan-XT) [reply dockers/docker-database/critical_processes.j2:3]: If there is no blank line here, all program:{{ redis_inst }} of critical_processes will be concentrated on one line, which will cause subsequent regular expression matching to fail.

File: dockers/docker-database/docker-database-init.sh (hunk 8 above)
Comment (Pan-XT) [reply dockers/docker-database/docker-database-init.sh:133]: Here are the necessary permissions for enable_multidb, see /etc/supervisor/conf.d/supervisord.conf.j2

File: dockers/docker-database/supervisord.conf.j2 (hunk 5 above)
Comment (Pan-XT) [reply dockers/docker-database/supervisord.conf.j2]: https://github.com/sonic-net/sonic-buildimage/pull/20545

File: dockers/docker-database/docker-database-init.sh (hunk 8 above)
Comment (liuh-80) [reply dockers/docker-database/docker-database-init.sh:133]: Hi @Pan-XT , can you add your comments to this file? that will help other understand the propose of this code.

File: dockers/docker-database/Dockerfile.j2 (hunk 3 above)
Comment (liuh-80) [reply dockers/docker-database/Dockerfile.j2:51]: #Closed

File: dockers/docker-database/critical_processes.j2 (hunk 9 above)
Comment (liuh-80) [reply dockers/docker-database/critical_processes.j2:3]: #Closed

File: dockers/docker-database/docker-database-init.sh (hunk 8 above)
Comment (Pan-XT) [reply dockers/docker-database/docker-database-init.sh:133]: ok，I add commnets “the Redis process is operating under the 'redis' user in supervisord and make redis user own /var/lib/$inst inside db container.”

Comment (liuh-80) [reply dockers/docker-database/docker-database-init.sh:133]: #Closed

File: dockers/docker-database/supervisord.conf.j2 (hunk 5 above)
Comment (liuh-80) [reply dockers/docker-database/supervisord.conf.j2]: #Closed

File: dockers/docker-database/multi_database_config.json.j2 (hunk 6 above)
Comment (Pan-XT) [reply dockers/docker-database/multi_database_config.json.j2:10]: https://github.com/sonic-net/sonic-buildimage/pull/5283

File: dockers/docker-database/multi_database_config.json.j2 (hunk 7 above)
Comment (Pan-XT) [reply dockers/docker-database/multi_database_config.json.j2:166]: Currently, the multi_database_config.json.j2 is consistent with the database_config.json.j2, except for the added content related to redis1,redis2,redis3,redis4. Defining a variable for all these dupe templates will be implemented in future modifications.

Comment (liuh-80) [reply dockers/docker-database/multi_database_config.json.j2:166]: #Closed

File: dockers/docker-database/multi_database_config.json.j2 (hunk 6 above)
Comment (liuh-80) [reply dockers/docker-database/multi_database_config.json.j2:10]: I search that PR, not find any code read "database_type"
This attribute is an optional attribute, and this file need be parsed by SonicDBConfig::parseDatabaseConfig
https://github.com/sonic-net/sonic-swss-common/blob/master/common/dbconnector.cpp

Otherwise, sonic-swss-common lib can't handle multi-DB, so please make sure sonic-swss-common also change to handle it.

Comment (Pan-XT) [reply dockers/docker-database/multi_database_config.json.j2:10]: In  /sonic-buildimage/dockers/docker-database/docker-database-init.sh line 53,

Comment (eddieruan-alibaba) [reply dockers/docker-database/multi_database_config.json.j2:10]: @liuh-80 

This field was added via 
https://github.com/sonic-net/sonic-buildimage/commit/9f08f88a0d1e45cdab3b1e5037a24908057e6129
by @Pterosaur

It is current SONiC behavior
https://github.com/sonic-net/sonic-buildimage/blob/master/dockers/docker-database/database_config.json.j2#L10

Can you help to check with him to see what the need for this field?

File: dockers/docker-database/critical_processes.j2 (hunk 9 above)
Comment (saiarcot895) [reply dockers/docker-database/critical_processes.j2:3]: Does changing `{%- endfor %}` to `{% endfor %}` work? That will make the intention a bit clearer.

File: dockers/docker-database/multi_database_config.json.j2 (hunk 6 above)
Comment (liuh-80) [reply dockers/docker-database/multi_database_config.json.j2:10]: #Closed, I will offline sync with Ze

File: dockers/docker-database/critical_processes.j2 (hunk 9 above)
Comment (Pan-XT) [reply dockers/docker-database/critical_processes.j2]: Changing {%- endfor %} to {% endfor %} works. I apply this change instead of adding blank line.

File: dockers/docker-database/multi_database_config.json.j2 (hunk 4 above)
Comment (oleksandrivantsiv) [reply dockers/docker-database/multi_database_config.json.j2]: #closed

**Comment by kperumalbfn** [comment]
@Pan-XT Could you add this build option ENABLE_MULTIDB in the multi DB HLD doc?

**Comment by kperumalbfn** [comment]
@Pan-XT Could you also sonic-mgmt tests for multi DB. Though this option is not enabled by default, will be good to have these tests in sonic-mgmt.

**Comment by eddieruan-alibaba** [comment]
> @Pan-XT Could you also sonic-mgmt tests for multi DB. Though this option is not enabled by default, will be good to have these tests in sonic-mgmt.

@kperumalbfn he replied it before. https://github.com/sonic-net/sonic-mgmt/pull/15084. We need to build a new image with this compile flag is enabled, then reuse existing test cases. No need to create new one since we don't expect any differences except performance gain.

**Comment by Pan-XT** [comment]
> @Pan-XT Could you also sonic-mgmt tests for multi DB. Though this option is not enabled by default, will be good to have these tests in sonic-mgmt.

I raise a PR https://github.com/sonic-net/SONiC/pull/1855

**Comment by Pan-XT** [comment]
> @Pan-XT Could you add this build option ENABLE_MULTIDB in the multi DB HLD doc?

I raise a PR https://github.com/sonic-net/SONiC/pull/1855

**Review by liuh-80 — APPROVED** [review]

File: dockers/docker-database/Dockerfile.j2 (hunk 3 above)
Comment (kperumalbfn) [reply dockers/docker-database/Dockerfile.j2:51]: @Pan-XT Could you update the HLD with some performance numbers with the multi DB testing?

Comment (kperumalbfn) [reply dockers/docker-database/Dockerfile.j2:51]: @Pan-XT Also provide a table with number of redis instances used during testing and what is the ideal number of instances used.

**Review by kperumalbfn — APPROVED** [review]
Thanks for the changes. Please update the HLD with the performance numbers.

**Comment by FengPan-Frank** [comment]
also some fix PR for multiple database process, otherwise redis instance cannot be launched correctly.

https://github.com/sonic-net/sonic-buildimage/pull/20545
https://github.com/sonic-net/sonic-buildimage/pull/20726

**Comment by Pan-XT** [comment]
> also some fix PR for multiple database process, otherwise redis instance cannot be launched correctly.
> 
> #20545 #20726

Thanks for the suggestion. These PRs are merged and I link them in description.
